state/upgrades.go

```diff
@@ -2961,3 +2961,88 @@ func AddCharmhubToModelConfig(pool *StatePool) error {
 		return false, nil
 	}))
 }
+
+// RollUpAndConvertOpenedPortDocuments replaces pre-2.9 per-machine, per-subnet
+// opened port documents with a single document that references port ranges
+// by endpoint names.
+//
+// This upgrade step exploits the fact that pre-2.9 controllers open ports
+// in all subnets. As a result, the opened ports collection will always
+// contain a single document with an empty subnet ID to indicate that the
+// port ranges apply to all subnets.
+func RollUpAndConvertOpenedPortDocuments(pool *StatePool) error {
+	const allEndpoints = ""
+	return errors.Trace(runForAllModelStates(pool, func(st *State) error {
+		col, closer := st.db().GetCollection(openedPortsC)
+		defer closer()
+
+		var oldDocs []upgrade.OldPortsDoc28
+		err := col.Find(nil).All(&oldDocs)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// Roll up per-subnet docs (there should actually only be one
+		// per machine) and emit txn ops for removing them.
```

Comment:
Maybe a check and warning if there do happen to be > 1 doc per machine, even though it should not happen

Reply:
Done